### ISSUE TITLE
Headers become blank space in gradle 2.0 and recyclerview 23.3

### ISSUE THREAD

**Issue by huteri** [timehop/sticky-headers-recyclerview#122]
Hi, I have issue with the new recyclerview 23.3.0, the usage is similar with the sample. Is there anybody facing similar issue?

![screenshot_20160421-160423](https://cloud.githubusercontent.com/assets/4598788/14703633/10a6a236-07db-11e6-8468-0d0dbcaef107.png)

**Comment by MatthieuFereyre** [comment]
Yes I do, and I manage to show my header by changing layout in my xml header file, but the header background is never painted (it is transparent and we can see items behind).

It also doesn't work with recyclerview-v7:23.2.1.

**Comment by MatthieuFereyre** [comment]
I managed to have my headers well displayed.

Check if your header layout has some height set to match_parent (set it to wrap_content).
Check if you set the background color of your header in the onBindHeaderViewHolder method.
It was transparent because in my layout it was just android:background="?panelBackgroundColor".

Hope this help.

**Comment by huteri** [comment]
Yeah, I set height to wrap content in header layout, and the background color is solid blue. Everything seems great, but can not make it work.

Here is the header item

```
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    android:orientation="horizontal"
    android:minHeight="8dp"
    android:gravity="center_vertical"
    android:background="@android:color/holo_blue_dark"
    android:paddingTop="10dp"
    android:paddingBottom="10dp"
    android:paddingRight="10dp"
    android:paddingLeft="20dp">

    <TextView
        android:id="@+id/date"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:textColor="@android:color/white"
        android:textSize="@dimen/text_size_14"
        android:text="{{HEADER}}"/>

</LinearLayout>
```

I don't set any background in the onBindViewHolder, the weird stuff here is that the problem only occurred on the first and second header, the rest is fine.

**Comment by chiraginexture** [comment]
i have same issue with     compile 'com.android.support:appcompat-v7:23.2.1'

every refresh adapter with  increasing header spacing. can u have solution??

**Comment by huteri** [comment]
Unfortunately, I have not yet found the solution, and eventually, I just use another library.

-- 
Huteri Manza

**Comment by koocbor** [comment]
Late to the game here - I was able to fix this issue by setting the header height to wrap_content AND the recycler view itself to be wrap_content.

**Comment by ivamluz** [comment]
Hi all,

Any other guess, other than changing height to wrap_content? The issue seems to happen more frequenly when my list has only a single item.

Also noticed it sometimes happen (intermittently) after removing items from the list.

Headers appear after I scroll the list.

**Comment by antonshkurenko** [comment]
I solved this calling invalidateHeaders in the post():

```
mInvalidate = new Runnable() {
    @Override
    public void run() {
         headersDecoration.invalidateHeaders();
    }
};

mAdapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
    @Override
    public void onChanged() {
         if (mInvalidate != null) {
              mRecyclerView.post(mInvalidate);
         }
    }

    @Override
    public void onItemRangeChanged(int positionStart, int itemCount) {
         if (mInvalidate != null) {
              mRecyclerView.post(mInvalidate);
         }
    }

    @Override
    public void onItemRangeInserted(int positionStart, int itemCount) {
         if (mInvalidate != null) {
              mRecyclerView.post(mInvalidate);
         }
    }

    @Override
    public void onItemRangeRemoved(int positionStart, int itemCount) {
         if (mInvalidate != null) {
              mRecyclerView.post(mInvalidate);
         }
    }
});
```

And also later I have bug with switching fragments and this, this hack solved my problem:

```
 //region Bug workaround
 // if I don't add this before next lines it blinks very shortly, but I see
 if (mInvalidate != null) {
     mRecyclerView.post(mInvalidate);
 }

 // small hack, somehow this library shows headers when I switch back to this fragment
 // only after scroll, that's why I scroll
 {
     mRecyclerView.smoothScrollBy(0, 1);
     mRecyclerView.smoothScrollBy(0, -1);
 }
 //endregion
```

**Comment by gabriel-hurtado** [comment]
Just so you know, tony's solution worked for me, but I had to add 
`
                mRecyclerView.invalidateItemDecorations();`

in the run()

**Comment by groomefyvik** [comment]
i am also facing the same issue ....any body found any solution .....
above solution didn't work for me

**Comment by cuberob** [comment]
Was having this issue when using the sticky adapter in combination with a recyclerview (height set to wrap_content) that was nested in a BottomSheetDialogFragment. 
Simply calling: 

```
mDecor = new StickyRecyclerHeadersDecoration(mAdapter);
mRecyclerView.addItemDecoration(mDecor);
mRecyclerView.post(new Runnable() {
            @Override
            public void run() {
                mDecor.invalidateHeaders();
            }
        });
```

After setting up the adapter seems to have fixed it for me.

**Comment by YousufSohail** [comment]
@huteri which other library you used?

**Comment by TadejZupancic** [comment]
I have debugged the library and found out, that the parent.getWidth() in getHeader function in HeaderViewCache sometimes returned 0 instead of the right width. But I've noticed, that the getMeasuredWidth() returned the right number every time.

So I've replaced parent.getWidth() with parent.getMeasuredWidth() and parent.getHeight() with parent.getMeasuredHeight() and now it seems to work fine. Previously the view height was too big almost every time and now I've run the application 50 times and it was showing headers fine all the time.

Since I don't know exactly how and when the views are measured, this might cause some other issues. If there's somebody that knows this things better than me, please comment.

**Comment by lanetteamtapan** [comment]
recyclerView.post(new Runnable() {
            @Override
            public void run() {
                recyclerView.invalidateItemDecorations();

                headersDecor.invalidateHeaders();
            }
        });

@gabriel-hurtado  thanks for advice this code worked for me.

**Comment by SoftwareAdvise** [comment]
Nobody have solved for me.. a library with a incredible bug..

**Comment by SoftwareAdvise** [comment]
Anybody have an alternative library for this scope????

**Comment by Andrej1A** [comment]
@SoftwareAdvise have you tried the solution from the last post of lanetteamtapan? You have to execute it in the `onResume()` method. It worked for me.

**Comment by PaolinoAngeletti** [comment]
It'd not worked for me . I'm using another library: 
compile 'org.zakariya.stickyheaders:stickyheaders:0.7.6'
https://github.com/ShamylZakariya/StickyHeaders

**Comment by liepang** [comment]
recyclerView.post(new Runnable() {
        @Override
        public void run() {
            recyclerView.invalidateItemDecorations();

            headersDecor.invalidateHeaders();
        }
    });

call this after finished bind the data

**Comment by rion18** [comment]
@TadejZupancic 's answer worked for me.

**Comment by TheKamranUllah** [comment]
> ```
>  recyclerView.post(new Runnable() {
>         @Override
>         public void run() {
>             recyclerView.invalidateItemDecorations();
> 
>             headersDecor.invalidateHeaders();
>         }
>     });
> ```
> 
> @gabriel-hurtado thanks for advice this code worked for me.



> ```
>  recyclerView.post(new Runnable() {
>         @Override
>         public void run() {
>             recyclerView.invalidateItemDecorations();
> 
>             headersDecor.invalidateHeaders();
>         }
>     });
> ```
> 
> @gabriel-hurtado thanks for advice this code worked for me.

I had an issu of extra spacing being generated by default when using the sticky adapter in combination with a recycler view (height set to wrap_content) that was nested in a BottomSheetDialogFragment inside ViewPager.
Calling the above snippet solved my problem.
